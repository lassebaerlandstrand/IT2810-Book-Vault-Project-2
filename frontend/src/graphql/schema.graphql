type Book {
  id: ID!
  title: String!
  authors: [String!]!
  publisher: String!
  genres: [String!]!
  coverImg: String!
}

type Query {
  books(limit: Int, offset: Int): [Book!]!
}
