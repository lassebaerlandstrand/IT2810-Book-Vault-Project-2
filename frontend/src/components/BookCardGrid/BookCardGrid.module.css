.link {
  text-decoration: none;
  color: inherit;

  & > * {
    transition: transform 0.15s ease-in-out;

    &:hover {
      transform: translateY(-5px);
      background-color: color-mix(in srgb, var(--mantine-color-dark-0) 25%, transparent);
    }
  }
}
