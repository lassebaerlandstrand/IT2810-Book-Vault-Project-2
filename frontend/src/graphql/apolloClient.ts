import { ApolloClient, InMemoryCache } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';

const client = new ApolloClient({
  uri: import.meta.env.VITE_GRAPHQL_ENDPOINT, // You can keep this URI for future use
  cache: new InMemoryCache({
    typePolicies: {
      Book: {
        keyFields: ['id'],
      },
      User: {
        keyFields: ['UUID'],
      },
    },
  }),
});

export { client, MockedProvider };
