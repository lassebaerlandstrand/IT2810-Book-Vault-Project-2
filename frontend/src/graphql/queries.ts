import { gql } from '@apollo/client';

export const GET_BOOKS = gql`
  query GetBooks($limit: Int, $offset: Int) {
    books(limit: $limit, offset: $offset) {
      id
      title
      authors
      publisher
      genres
      coverImg
    }
  }
`;
