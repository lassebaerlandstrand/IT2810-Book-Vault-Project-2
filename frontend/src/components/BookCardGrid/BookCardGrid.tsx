import { ApolloError } from '@apollo/client';
import { IconBookOff } from '@tabler/icons-react';
import { Link } from 'react-router-dom';
import { Grid, Group, Stack, Text } from '@mantine/core';
import { Book } from '@/generated/graphql';
import BookCard from '../BookCard/BookCard';
import BookListViewCard from '../BookListViewCard/BookListViewCard';
import Loading from '../Loading/Loading';
import styles from './BookCardGrid.module.css';

type BookCardGridProps = {
  books:
    | Pick<
        Book,
        'id' | 'title' | 'coverImg' | 'rating' | 'authors' | 'numRatings' | 'genres' | 'description'
      >[]
    | undefined;
  loading: boolean;
  error: ApolloError | undefined;
  viewType?: 'grid' | 'list';
};

/**
 * Displays a grid of book cards.
 */
const BookCardGrid = ({ books, loading, error, viewType = 'grid' }: BookCardGridProps) => {
  if (loading) {
    return <Loading />;
  }

  if (error || !books) {
    return (
      <Group justify="center" align="flex-start" className={styles.noResultWrapper}>
        <Group align="center">
          <IconBookOff />
          <Text size="xl" fw={500} my="xl">
            Error fetching books
          </Text>
        </Group>
      </Group>
    );
  }

  if (books.length === 0) {
    return (
      <Group justify="center" align="flex-start" className={styles.noResultWrapper}>
        <Group align="center">
          <IconBookOff />
          <Text size="xl" fw={500} my="xl">
            No books found
          </Text>
        </Group>
      </Group>
    );
  }

  if (viewType === 'list') {
    return (
      <Stack gap={20}>
        {books.map((book) => (
          <Link to={`/book/${book.id}`} className={styles.link} key={book.id}>
            <BookListViewCard book={book} />
          </Link>
        ))}
      </Stack>
    );
  }

  return (
    <>
      <Grid justify="center" display="inline-grid">
        {books.map((book) => (
          <Grid.Col key={book.id} span="content" className={styles.column}>
            <Link to={`/book/${book.id}`} className={styles.link}>
              <BookCard book={book} />
            </Link>
          </Grid.Col>
        ))}
      </Grid>
    </>
  );
};

export default BookCardGrid;
