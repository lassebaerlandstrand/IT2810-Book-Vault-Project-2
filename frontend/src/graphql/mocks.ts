import { fetchAuthors, fetchBooks, fetchGenres, fetchPublishers } from '../api/dummyApi';
import { GET_AUTHORS, GET_BOOKS, GET_GENRES, GET_PUBLISHERS } from './queries';

export const mocks = [
  {
    request: {
      query: GET_BOOKS,
      variables: {
        limit: 10,
        offset: 0,
      },
    },
    result: {
      data: {
        books: fetchBooks().slice(0, 10),
      },
    },
  },
  {
    request: {
      query: GET_AUTHORS,
      variables: {
        limit: 10,
        offset: 0,
      },
    },
    result: {
      data: {
        authors: fetchAuthors().slice(0, 10),
      },
    },
  },
  {
    request: {
      query: GET_PUBLISHERS,
      variables: {
        limit: 10,
        offset: 0,
      },
    },
    result: {
      data: {
        publishers: fetchPublishers().slice(0, 10),
      },
    },
  },
  {
    request: {
      query: GET_GENRES,
      variables: {
        limit: 10,
        offset: 0,
      },
    },
    result: {
      data: {
        genres: fetchGenres().slice(0, 10),
      },
    },
  },
];
