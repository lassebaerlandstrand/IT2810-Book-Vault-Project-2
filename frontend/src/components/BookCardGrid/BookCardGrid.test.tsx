import { render } from '@test-utils';
import { MemoryRouter } from 'react-router-dom';
import { Book } from '@/generated/graphql';
import BookCardGrid from './BookCardGrid';

const dummyBook: Book = {
  __typename: 'Book',
  authors: ['Author'],
  coverImg: 'https://placehold.co/200x300?text=Cover%20image%20for%20book',
  genres: ['Genre'],
  id: '1',
  publisher: 'Publisher',
  title: 'Title',
  bookFormat: '',
  characters: [],
  description: '',
  isbn: '',
  language: '',
  pages: 0,
  publishDate: '',
  rating: 0,
  series: '',
};

const dummyBooks: Book[] = Array.from({ length: 10 }, (_, index) => ({
  ...dummyBook,
  id: (index + 1).toString(),
}));

describe('Bookcard component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <BookCardGrid books={dummyBooks} />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
