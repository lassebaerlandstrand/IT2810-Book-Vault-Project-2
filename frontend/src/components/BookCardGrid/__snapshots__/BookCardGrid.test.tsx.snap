// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Bookcard component > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <style
    data-mantine-styles="inline"
  >
    .__m__-r0{--grid-gutter:var(--mantine-spacing-md);}
  </style>
  <div
    class="m_410352e9 mantine-Grid-root __m__-r0"
    style="margin-block: var(--mantine-spacing-xl);"
  >
    <div
      class="m_dee7bd2f mantine-Grid-inner"
    >
      <style
        data-mantine-styles="inline"
      >
        .__m__-r2{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-r2{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-r2{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-r2{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-r2{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-r2"
      >
        <a
          class="_link_7d2061"
          href="/book/1"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-ra{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-ra{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-ra{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-ra{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-ra{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-ra"
      >
        <a
          class="_link_7d2061"
          href="/book/2"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-ri{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-ri{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-ri{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-ri{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-ri{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-ri"
      >
        <a
          class="_link_7d2061"
          href="/book/3"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-rq{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-rq{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-rq{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-rq{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-rq{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-rq"
      >
        <a
          class="_link_7d2061"
          href="/book/4"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-r12{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-r12{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-r12{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-r12{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-r12{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-r12"
      >
        <a
          class="_link_7d2061"
          href="/book/5"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-r1a{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-r1a{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-r1a{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-r1a{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-r1a{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-r1a"
      >
        <a
          class="_link_7d2061"
          href="/book/6"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-r1i{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-r1i{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-r1i{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-r1i{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-r1i{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-r1i"
      >
        <a
          class="_link_7d2061"
          href="/book/7"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-r1q{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-r1q{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-r1q{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-r1q{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-r1q{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-r1q"
      >
        <a
          class="_link_7d2061"
          href="/book/8"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-r22{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-r22{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-r22{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-r22{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-r22{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-r22"
      >
        <a
          class="_link_7d2061"
          href="/book/9"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
      <style
        data-mantine-styles="inline"
      >
        .__m__-r2a{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 24em){.__m__-r2a{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 30em){.__m__-r2a{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}@media(min-width: 48em){.__m__-r2a{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}@media(min-width: 64em){.__m__-r2a{--col-flex-grow:auto;--col-flex-basis:25%;--col-max-width:25%;}}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-r2a"
      >
        <a
          class="_link_7d2061"
          href="/book/10"
        >
          <div
            class="_card_0962fe m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
            style="--paper-radius: var(--mantine-radius-lg); margin: auto;"
          >
            <div
              class="m_599a2148 mantine-Card-section"
              data-first-section="true"
            >
              <img
                alt="Cover image for A name"
                class="m_9e117634 mantine-Image-root"
                src="https://placehold.co/200x300?text=Cover%20image%20for%20book"
                style="--image-radius: var(--mantine-radius-lg); --image-object-fit: contain; margin: auto; max-width: 100%; max-height: 100%;"
              />
            </div>
            <div
              class="m_599a2148 mantine-Card-section _textSection_0962fe"
              data-last-section="true"
            >
              <h4
                class="mantine-focus-auto _bookTitle_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                style="--text-line-clamp: 2; font-weight: 600;"
                title="A name"
              >
                A name
              </h4>
              <p
                class="mantine-focus-auto _bookAuthor_0962fe m_b6d8b162 mantine-Text-root"
                data-line-clamp="true"
                data-size="xs"
                style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); --text-line-clamp: 2;"
                title="An author"
              >
                An author 
              </p>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
